File: src/main/java/org/radarcns/stream/empatica/E4InterBeatIntervalStream.java
/*
 * Copyright 2017 King's College London and The Hyve
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.radarcns.stream.empatica;

import java.io.IOException;
import javax.annotation.Nonnull;
import org.apache.kafka.streams.kstream.KStream;
import org.apache.kafka.streams.kstream.TimeWindows;
import org.radarcns.aggregator.DoubleAggregator;
import org.radarcns.config.KafkaProperty;
import org.radarcns.empatica.EmpaticaE4InterBeatInterval;
import org.radarcns.key.MeasurementKey;
import org.radarcns.key.WindowedKey;
import org.radarcns.stream.StreamWorker;
import org.radarcns.stream.StreamMaster;
import org.radarcns.stream.collector.DoubleValueCollector;
import org.radarcns.util.RadarSingletonFactory;
import org.radarcns.util.RadarUtilities;
import org.radarcns.util.serde.RadarSerdes;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Definition of Kafka Stream for aggregating Inter Beat Interval values collected by Empatica E4
 */
public class E4InterBeatIntervalStream extends
        StreamWorker<MeasurementKey, EmpaticaE4InterBeatInterval> {
    private static final Logger log = LoggerFactory.getLogger(E4InterBeatIntervalStream.class);
    private final RadarUtilities utilities = RadarSingletonFactory.getRadarUtilities();

    public E4InterBeatIntervalStream(String clientId, int numThread, StreamMaster master,
            KafkaProperty kafkaProperties) {
        super(E4Streams.getInstance().getInterBeatIntervalStream(),
                clientId, numThread, master, kafkaProperties, log);
    }

    @Override
    protected KStream<WindowedKey, DoubleAggregator> setStream(@Nonnull KStream<MeasurementKey, EmpaticaE4InterBeatInterval> kstream)
            throws IOException {
        return kstream.groupByKey()
                .aggregate(
                    DoubleValueCollector::new,
                    (k, v, valueCollector) -> valueCollector.add(extractValue(v)),
                    TimeWindows.of(10 * 1000L),
                    RadarSerdes.getInstance().getDoubleCollector(),
                    getStreamDefinition().getStateStoreName())
                .toStream()
                .map(utilities::collectorToAvro);
    }

    private Float extractValue(EmpaticaE4InterBeatInterval record) {
        incrementMonitor();
        return record.getInterBeatInterval();
    }
}

File: src/main/kotlin/org/radarbase/kotlin/stream/empatica/E4InterBeatIntervalStream.kt
package org.radarbase.kotlin.stream.empatica

import org.apache.kafka.streams.kstream.KStream
import org.radarbase.config.RadarPropertyHandler
import org.radarbase.kotlin.stream.SensorStreamWorker
import org.radarbase.kotlin.stream.StreamDefinition
import org.radarcns.kafka.AggregateKey
import org.radarcns.kafka.ObservationKey
import org.radarcns.passive.empatica.EmpaticaE4InterBeatInterval
import org.radarcns.stream.aggregator.NumericAggregate

/**
 * Definition of Kafka Stream for aggregating Inter Beat Interval values collected by Empatica E4.
 */
class E4InterBeatIntervalStream : SensorStreamWorker<ObservationKey, EmpaticaE4InterBeatInterval>() {
    override fun initialize() {
        defineWindowedSensorStream("android_empatica_e4_inter_beat_interval")
        config.setDefaultPriority(RadarPropertyHandler.Priority.LOW)
    }

    override fun implementStream(
        definition: StreamDefinition,
        kstream: KStream<ObservationKey, EmpaticaE4InterBeatInterval>
    ): KStream<AggregateKey, NumericAggregate> {
        return aggregateNumeric(
            definition, kstream, "interBeatInterval",
            EmpaticaE4InterBeatInterval.getClassSchema()
        )
    }
}
